keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -56,30 +67,156 @@ class NotificationEventListener(
-            val handler = handlers.find { hand ->
-              hand.type == notification.type
-            } as? SlackNotificationHandler<SlackUnpinnedNotification>
-            handler?.let {
-              it.sendMessage(SlackUnpinnedNotification(
-                channel = notificationConfig.address,
-                latestArtifact = latestArtifact,
-                pinnedVersion = pinnedEnvironment!!.version,
-                application = config.application,
-                time = clock.instant(),
-                user = user
-              ))
-            }
-
-          } else {
-            log.debug("no pinned artifacts exists for application ${config.application} and environment $targetEnvironment")
+      if (pinnedEnvironment == null) {
+        log.warn("no pinned artifacts exists for application ${config.application} and environment $targetEnvironment")
+        return
+      }
+
+      val latestApprovedArtifactVersion = repository.latestVersionApprovedIn(config, pinnedEnvironment!!.artifact, targetEnvironment)
+      if (latestApprovedArtifactVersion == null) {
+        log.warn("latestApprovedArtifactVersion is null for application ${config.application}, env $targetEnvironment. Can't send UnpinnedNotification")
+        return
+      }
+
+      val latestArtifact = repository.getArtifactVersion(pinnedEnvironment!!.artifact, latestApprovedArtifactVersion, null)
+      val pinnedArtifact = repository.getArtifactVersion(pinnedEnvironment!!.artifact, pinnedEnvironment!!.version, null)
+
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackUnpinnedNotification> ?: return
+
+      config.environments.forEach { environment ->
+        environment.notifications.filter {
+          it.type == NotificationType.slack
+        }.onEach {
+          handler.sendMessage(SlackUnpinnedNotification(
+            channel = it.address,
+            latestArtifact = latestArtifact,
+            pinnedArtifact = pinnedArtifact,
+            application = config.application,
+            time = clock.instant(),
+            user = user,
+            targetEnvironment = targetEnvironment
+          ))
+        }
+      }
+    }
+  }
+
+  @EventListener(MarkAsBadNotification::class)
+  fun onMarkAsBadNotification(notification: MarkAsBadNotification) {
+    with(notification) {
+      val deliveryArtifact = config.artifacts.find {
+        it.reference == veto.reference
+      } ?: return
+
+      val vetoedArtifact = repository.getArtifactVersion(deliveryArtifact, veto.version, null)
+      if (vetoedArtifact == null) {
+        log.warn("vetoedArtifact is null for application ${config.application}. Can't send MarkAsBadNotification notification")
+        return
+      }
+
+      val handler = handlers.find {
+        it.type == notification.type
+      } as? SlackNotificationHandler<SlackMarkAsBadNotification> ?: return
```

Comment:
Debug log before returning?

Reply:
P.S. One more reason for that `supporting` function: you can log in a single place when the handler is not found.

Reply:
done!